[5.2] Bug 1837680: Add Message-ID header when missing at send

mrenvoize opened the pull request:
#### Details
Detect whether a Message-ID header has been included in the message passed to MessageToMTA and generate one if it's missing.

GMail and other modern mail services now expect a Message-ID as part of the message else they mark them as spam or drop them entirely.

#### Additional info
* [bug 1837680](https://bugzilla.mozilla.org/show_bug.cgi?id=1837680)

mrenvoize committed:
Bug 1837680: Add Message-ID header when missing at send

Detect whether a Message-ID header has been included in the message
passed to MessageToMTA and generate one if it's missing.

GMail and other modern mail services now expect a Message-ID as part of
the message else they mark them as spam or drop them entirely.

mrenvoize commented:
I would have liked to make the new build_message_id sub a little more generic and use it for both build_thread_marker and MessageToMTA, but I felt we could do that at a later stage.

I'm also not entirely sure about the user-id inclusion here.. I'm not close enough to the codebase with regards to how Bugzilla->user works yet to know whether it will get the right user in this case.. I also wasn't sure how I could get back to the userid within MessateToMTA.. I wondered about trying to extract the 'To' address and perform a lookup of users on that?

Anyway, here's a first draft for review.  Unit tests also to follow once I've had some initial feedback.

mrenvoize committed:
Bug 183780: Add POD for new method

justdave commented on Bugzilla/Mailer.pm:
```diff
@@ -223,6 +227,26 @@ sub build_thread_marker {
+
+  my $sitespec = '@' . Bugzilla->params->{'urlbase'};
+  $sitespec =~ s/:\/\//\./;               # Make the protocol look like part of the domain
+  $sitespec =~ s/^([^:\/]+):(\d+)/$1/;    # Remove a port number, to relocate
```
This does not appear to adequately sanitize the potential values of urlbase, since it might contain a path after the domain name and / is also an illegal character after the @ in the message ID.

for example, if urlbase = `https://bugs.example.org/bugzilla/` then after these regexps, sitespec becomes roughly `@https.bugs.example.org/bugzilla/`

This is apparently an issue in the existing build_thread_marker that you probably copied this logic from, so would need to be dealt with in both places. Since the whole point of this bug is fixing acceptance and places that are strict about Message-IDs, that probably should be included in this patch.

justdave commented on Bugzilla/Mailer.pm:
```diff
@@ -223,6 +227,26 @@ sub build_thread_marker {
+
+  if (!defined $user_id) {
+    $user_id = Bugzilla->user->id;
+  }
```
There are several places this MessageToMTA gets called from where there will be no user, and there's also a few places where the mail being sent may be irrelevant to the current logged in user. It might be safer to leave this blank or use a static string instead if it's not passed in, to avoid tagging a user to something they weren't responsible for. The random string added later should take care of it still being unique.

justdave commented on Bugzilla/Mailer.pm:
```diff
@@ -223,6 +227,26 @@ sub build_thread_marker {
+  $sitespec =~ s/:\/\//\./;               # Make the protocol look like part of the domain
+  $sitespec =~ s/^([^:\/]+):(\d+)/$1/;    # Remove a port number, to relocate
+  if ($2) {
+    $sitespec = "-$2$sitespec";           # Put the port number back in, before the '@'
```
My syntax checker is warning me that $2 may contain values from a previous regexp match attempt if this one fails to match anything, since Perl doesn't clear them in between. It's suggesting the code should check if the regexp succeeded or not before depending on a match position variable to work.

justdave reviewed (changes requested):
The issues mentioned here apply to the Harmony PR as well ([bugzilla/harmony#146](https://github.com/bugzilla/harmony/pull/146))

mrenvoize committed:
Bug 1837680: Fix Message-ID sitespec sanitization and user fallback

- Strip path components from urlbase when building sitespec, as '/' is
  illegal after '@' in a Message-ID (affects both build_thread_marker
  and build_message_id)
- Fix stale $2 capture variable risk by conditioning port relocation on
  the substitution succeeding, in both functions
- Avoid falling back to Bugzilla->user->id in build_message_id, which
  is called from contexts with no logged-in user; random bits ensure
  uniqueness without a user identifier
- Omit the user_id segment entirely when absent to avoid a double-dash
  in the Message-ID local-part

mrenvoize committed:
Bug 1837680: Add unit tests for build_message_id and build_thread_marker

Test the sitespec transformation for four urlbase forms (plain, with
path, with port, with port+path) to guard against regressions in the
'/' stripping and port relocation logic. Also covers user_id handling
in build_message_id (present, absent, no double-dash).

justdave commented on Bugzilla/Mailer.pm:
```diff
@@ -264,9 +264,9 @@ sub build_thread_marker {
-  $sitespec =~ s/^([^:\/]+):(\d+)/$1/;    # Remove a port number, to relocate
-  if ($2) {
-    $sitespec = "-$2$sitespec";    # Put the port number back in, before the '@'
+  $sitespec =~ s/\/.*$//;      # Strip path component — / is illegal after @ in Message-ID
```
I think we actually want the path component because you might have more than one bugzilla on the same host in different subdirectories. But the / should probably be changed to a . or an _ or something. Actually, looking how we're handling the port number, maybe we should reposition the path component in front as well, so the actual domain name is still last in the sitespec.

mrenvoize replied:
Good catch, follow-up pushed

mrenvoize committed:
Bug 1837680: Relocate urlbase path component instead of discarding it

Stripping the path from urlbase loses information: two independent
Bugzilla installs hosted in different subdirectories of the same
domain (e.g. /team-a/ and /team-b/) would produce identical sitespecs,
letting mail clients that group by References/In-Reply-To/Message-ID
cross-thread unrelated bugs from separate installations.

Relocate the path in front of the domain instead, the same way the
port number is already relocated, sanitizing any internal '/' since
it's illegal in a domain. Updates build_thread_marker and
build_message_id identically, plus the corresponding tests.

justdave reviewed (approved):
Fixes look good.  I wasn't expecting subdir and port to end up on the front side of the @ but that's legal and works for me. :-)
